pallets/identity/src/keys.rs

```diff
@@ -189,22 +190,44 @@ impl<T: Config> Module<T> {
         Self::accept_auth_with(&signer, rotation_auth_id, |data, target_did| {
             // Ensure Authorization is a `RotatePrimaryKey`.
             extract_auth!(data, RotatePrimaryKey);
-            Self::unsafe_primary_key_rotation(sender, target_did, optional_cdd_auth_id)
+            Self::common_rotate_primary_key(target_did, sender, None, optional_cdd_auth_id)
         })
     }
 
-    /// Processes primary key rotation.
-    pub fn unsafe_primary_key_rotation(
-        sender: T::AccountId,
-        rotation_for_did: IdentityId,
+    // Sets the new primariy key, optionally removes it as a secondary key if it is one,
+    // and optionally checks the cdd auth.
+    pub fn common_rotate_primary_key(
+        target_did: IdentityId,
+        new_primary_key: T::AccountId,
+        new_permissions: Option<Permissions>,
         optional_cdd_auth_id: Option<u64>,
     ) -> DispatchResult {
+        let mut record = <DidRecords<T>>::get(target_did);
+
+        let is_linked = <KeyToIdentityIds<T>>::contains_key(&new_primary_key);
+        let is_secondary_key =
+            is_linked && <KeyToIdentityIds<T>>::get(&new_primary_key) == target_did;
+        let is_primary_key = record.primary_key == new_primary_key;
+        let is_multisig_signer = T::MultiSig::is_signer(&new_primary_key);
+
+        ensure!(
+            (!is_linked || is_secondary_key) && !is_primary_key && !is_multisig_signer,
+            Error::<T>::AlreadyLinked
+        );
+
+        let old_primary_key = record.primary_key.clone();
+
+        if new_permissions.is_none() {
```

Comment:
Why do we only need this check if the `new_permissions` is None? cc @Neopallium

Reply:
That is for the cases where the old primary will be left unlinked.  `new_permissions` will only be `Some(Perms)` if the old primary key will be moved to a secondary key with those permissions.

Reply:
Here we check for none, because we also need to check if the old primary key can be unlinked.

Reply:
Yeah - thanks - realised that as I got further down the review ;-)